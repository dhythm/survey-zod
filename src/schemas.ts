import { z } from "zod";

const passwordSchema = z
  .string()
  .min(8)
  .max(16)
  .refine((data) => data.match(/[a-z]/g) && data.match(/[A-Z]/g), {
    message: "password must include capital letter and small letter.",
  })
  .refine((data) => data.match(/[0-9]/g) && data.match(/[a-zA-Z]/g), {
    message: "password must include alphanumeric.",
  })
  .refine(
    (data) =>
      // https://javascript.info/regexp-escaping
      // https://javascript.info/regexp-character-sets-and-ranges#escaping-in-
      data.match(/[!"#\$%&'\(\)\*\+\-\.,\/:;<=>\?@\[\\\]\^_`{\|}~]/g),
    {
      message: "password must include special character.",
    }
  );

export const loginSchema = z.object({
  email: z.string().email(),
  password: passwordSchema,
});

export const changePasswordSchema = z
  .object({
    oldPassword: z.string().nonempty(),
    newPassword: passwordSchema,
    confirmationPassword: passwordSchema,
  })
  .refine(
    (data) =>
      !data.confirmationPassword ||
      data.newPassword === data.confirmationPassword,
    {
      path: ["confirmationPassword"],
      message: "confirmationPassword does not match.",
    }
  );

const teacherBaseSchema = z.object({ students: z.array(z.string()) });
const hasIdSchema = z.object({ id: z.string() });
const teacherSchema = teacherBaseSchema.merge(hasIdSchema);
const alternativeTeacherSchema = teacherBaseSchema.extend(hasIdSchema.shape);
const noIdTeacherSchema = teacherSchema.omit({ id: true });
const alternativeNoIdTeacherSchema = teacherSchema.pick({ students: true });
const partialTeacherSchema = teacherSchema.partial();
const optionalIdTeacherSchema = teacherSchema.partial({ id: true });

const userSchema = z.object({
  username: z.string(),
  location: z.object({
    latitude: z.number(),
    longitude: z.number(),
  }),
  strings: z.array(z.object({ value: z.string() })),
});
const deepPartialUserSchema = userSchema.deepPartial();

const personSchema = z.object({ name: z.string() });
const employeeSchema = z.object({ role: z.string() });
const employedPersonScheme = z.intersection(personSchema, employeeSchema);
// equivalent to: personSchema.and(employeeSchema)
// CAVEAT:
// Though in many cases, it is recommended to use A.merge(B) to merge two objects.
// The.merge method returns a new ZodObject instance, whereas A.and(B) returns a less useful ZodIntersection instance that lacks common object methods like pick and omit.
// https://github.com/colinhacks/zod#intersections

interface Category {
  name: string;
  subCategories: Category[];
}
const categorySchema: z.ZodType<Category> = z.lazy(() =>
  z.object({ name: z.string(), subCategories: z.array(categorySchema) })
);

// https://github.com/colinhacks/zod#json-type
const literalSchema = z.union([z.string(), z.number(), z.boolean(), z.null()]);
type Json = z.infer<typeof literalSchema> | { [key: string]: Json } | Json[];
export const jsonSchema: z.ZodType<Json> = z.lazy(() =>
  z.union([literalSchema, z.array(jsonSchema), z.record(jsonSchema)])
);

function makeSchemaOptional<T extends z.ZodTypeAny>(schema: T) {
  return schema.optional();
}
function makeStringSchemaOptional<T extends z.ZodType<string>>(schema: T) {
  return schema.optional();
}
const arg = makeSchemaOptional(z.string());
makeStringSchemaOptional(z.string());
// makeStringSchemaOptional(z.number());

// const nameAndEmailSchema = z.object({
//   name: z.string(),
//   email: z.string().email(),
// });
const nameAndEmailSchema = z
  .object({
    name: z.string(),
    email: z.string().email(),
  })
  .or(
    z.object({
      name: z.literal(""),
      email: z.literal(""),
    })
  );
const alternativeNameAndEmailSchema = z.union([
  z.object({
    name: z.string(),
    email: z.string().email(),
  }),
  z.object({
    name: z.literal(""),
    email: z.literal(""),
  }),
]);

export const nameAndEmailArraySchema = z.array(nameAndEmailSchema);
