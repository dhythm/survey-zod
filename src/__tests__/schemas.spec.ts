import { z, ZodError } from "zod";
import { changePasswordSchema, loginSchema } from "../schemas";

describe("zod", () => {
  describe("loginSchema", () => {
    it("should be passed", () => {
      expect(() =>
        loginSchema.parse({
          email: "test@example.com",
          password: "Password1!",
        })
      ).not.toThrowError();
    });
    it("should be failed", () => {
      try {
        loginSchema.parse({
          email: "",
          password:
            "12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901",
        });
      } catch (err) {
        expect(err as ZodError).toMatchSnapshot();
      }
    });
  });

  describe("changePasswordSchema", () => {
    it("should be passed", () => {
      expect(() =>
        changePasswordSchema.parse({
          oldPassword: "Password1!",
          newPassword: "Password2!",
          confirmationPassword: "Password2!",
        })
      ).not.toThrowError();
    });
    it("should be failed", () => {
      try {
        changePasswordSchema.parse({
          oldPassword: "",
          newPassword: "Password2!",
          confirmationPassword: "Password1!",
        });
      } catch (err) {
        expect(err as ZodError).toMatchSnapshot();
      }
    });
  });

  describe("date", () => {
    const dateSchema = z.preprocess((arg) => {
      if (typeof arg === "string" || arg instanceof Date) return new Date(arg);
    }, z.date());

    it("should be passed", () => {
      expect(() => dateSchema.parse(new Date("1/12/22"))).not.toThrowError();
      expect(() =>
        dateSchema.parse("2022-01-12T00:00:00.000Z")
      ).not.toThrowError();
    });
  });

  describe("enum", () => {
    // `as const` must be needed, otherwise z.enum will show error.
    const values = ["Salmon", "Tuna", "Trout"] as const;
    const enumSchema = z.enum(values);

    it("should be passed", () => {
      expect(enumSchema.options).toEqual(["Salmon", "Tuna", "Trout"]);
      expect(enumSchema.enum.Salmon).toBe("Salmon");
    });
  });

  describe("discriminatedUnion", () => {
    const discriminatedUnionSchema = z.discriminatedUnion("type", [
      z.object({ type: z.literal("a"), a: z.string() }),
      z.object({ type: z.literal("b"), b: z.string() }),
    ]);

    it("should be passed", () => {
      expect(() =>
        discriminatedUnionSchema.parse({ type: "a", a: "abc" })
      ).not.toThrowError();
    });
    it("should be failed", () => {
      try {
        discriminatedUnionSchema.parse({
          type: "a",
          b: "abc",
        });
      } catch (err) {
        expect(err as ZodError).toMatchSnapshot();
      }
      try {
        discriminatedUnionSchema.parse({
          type: "b",
          a: "abc",
        });
      } catch (err) {
        expect(err as ZodError).toMatchSnapshot();
      }
    });
  });

  describe("promise", () => {
    const promiseSchema = z.promise(z.number());

    it("should be passed", async () => {
      expect(await promiseSchema.parse(Promise.resolve(3.14))).toBe(3.14);
      expect(promiseSchema.parse(Promise.resolve(3.14))).resolves.toBe(3.14);
    });
    it("should be failed", () => {
      expect(
        promiseSchema.parse(Promise.resolve("tuna"))
      ).rejects.toThrowError();
      expect(() => promiseSchema.parse(3.14)).toThrowError();
    });
  });

  describe("function", () => {
    const functionSchema = z
      .function()
      .args(z.string())
      .returns(z.number())
      .implement((x) => x.trim().length);

    it("should be passed", () => {
      expect(functionSchema("sandwich")).toBe(8);
      expect(functionSchema(" asdf ")).toBe(4);
    });
  });

  describe("preprocess", () => {
    // Typically Zod operates under a "parse then transform" paradigm.
    // Zod validates the input first, then passes it through a chain of transformation functions.
    const preprocessSchema = z.preprocess((v) => String(v), z.string());

    it("should be passed", () => {
      expect(preprocessSchema.parse(8)).toBe("8");
    });
  });

  describe("schema methods", () => {
    it("should be passed", async () => {
      expect(
        z
          .string()
          .refine(async (val) => val.length > 20)
          .parseAsync("hello, folks! it is awesome")
      ).resolves.toBe("hello, folks! it is awesome");
    });
    it("should be failed", () => {
      expect(
        z
          .string()
          .refine(async (val) => val.length > 20)
          .parseAsync("hello")
      ).rejects.toThrowError();
    });
  });

  describe("refine and superRefine", () => {
    const refineSchema = z
      .string()
      .transform((val) => val.length)
      .refine((val) => val > 25);
    const superRefineSchema = z.array(z.string()).superRefine((val, ctx) => {
      if (val.length > 3) {
        ctx.addIssue({
          code: z.ZodIssueCode.too_big,
          maximum: 3,
          type: "array",
          inclusive: true,
          message: "Too many items ðŸ˜¡",
        });
      }

      if (val.length !== new Set(val).size) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: `No duplicated allowed.`,
        });
      }
    });

    it("should be passed", () => {
      expect(refineSchema.parse("abcdefghijklmnopqrstuvwxyz")).toBe(26);
      expect(() => superRefineSchema.parse(["a", "b"])).not.toThrowError();
    });
    it("should be failed", () => {
      expect(() => refineSchema.parse("abc")).toThrowError();
      try {
        superRefineSchema.parse(["a", "b", "C", "d"]);
      } catch (err) {
        expect(err as ZodError).toMatchSnapshot();
      }
      try {
        superRefineSchema.parse(["a", "a", "b"]);
      } catch (err) {
        expect(err as ZodError).toMatchSnapshot();
      }
    });
  });

  describe("transform", () => {
    // CAVEAT:
    // Transform functions must not throw.
    // Make sure to use refinements before the transform or addIssue within the transform to make sure the input can be parsed by the transform.
    // https://github.com/colinhacks/zod#transform
    const transformValidationSchema = z.string().transform((val, ctx) => {
      const parsed = parseInt(val);
      if (isNaN(parsed)) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "Not a number",
        });
      }
      return parsed;
    });
    const transformRefineSchema = z
      .string()
      .transform((val) => val.toUpperCase())
      .refine((val) => val.length < 15)
      .transform((val) => `Hello ${val}`)
      .refine((val) => val.indexOf("!") === -1);

    it("should be passed", () => {
      expect(
        z
          .string()
          .email()
          .transform((val) => val.split("@")[1])
          .parse("test@example.com")
      ).toBe("example.com");
      expect(transformValidationSchema.parse("16")).toBe(16);
      expect(transformRefineSchema.parse("World")).toBe("Hello WORLD");
    });

    it("should be failed", () => {
      try {
        expect(transformValidationSchema.parse("foo"));
      } catch (err) {
        expect(err as ZodError).toMatchSnapshot();
      }
      expect(() => transformRefineSchema.parse("World!")).toThrowError();
    });
  });

  describe("Error handling and error formatting", () => {
    it("should be passed", () => {});
    it("should be failed", () => {
      let data = z.object({ name: z.string() }).safeParse({ name: 12 });
      if (!data.success) {
        expect(data.error.issues).toMatchSnapshot();
        expect(data.error.format()).toMatchSnapshot();
      }
    });
  });

  describe("", () => {
    it("should be passed", () => {});
    it("should be failed", () => {});
  });
});
