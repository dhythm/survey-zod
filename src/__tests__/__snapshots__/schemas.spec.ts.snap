// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zod Error handling and error formatting should be failed 1`] = `
Array [
  Object {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected string, received number",
    "path": Array [
      "name",
    ],
    "received": "number",
  },
]
`;

exports[`zod Error handling and error formatting should be failed 2`] = `
Object {
  "_errors": Array [],
  "name": Object {
    "_errors": Array [
      "Expected string, received number",
    ],
  },
}
`;

exports[`zod changePasswordSchema should be failed 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "oldPassword"
    ]
  },
  {
    "code": "custom",
    "path": [
      "confirmationPassword"
    ],
    "message": "confirmationPassword does not match."
  }
]]
`;

exports[`zod discriminatedUnion should be failed 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "a"
    ],
    "message": "Required"
  }
]]
`;

exports[`zod discriminatedUnion should be failed 2`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "b"
    ],
    "message": "Required"
  }
]]
`;

exports[`zod loginSchema should be failed 1`] = `
[ZodError: [
  {
    "validation": "email",
    "code": "invalid_string",
    "message": "Invalid email",
    "path": [
      "email"
    ]
  },
  {
    "code": "too_big",
    "maximum": 16,
    "type": "string",
    "inclusive": true,
    "message": "String must contain at most 16 character(s)",
    "path": [
      "password"
    ]
  },
  {
    "code": "custom",
    "message": "password must include capital letter and small letter.",
    "path": [
      "password"
    ]
  },
  {
    "code": "custom",
    "message": "password must include alphanumeric.",
    "path": [
      "password"
    ]
  },
  {
    "code": "custom",
    "message": "password must include special character.",
    "path": [
      "password"
    ]
  }
]]
`;

exports[`zod refine and superRefine should be failed 1`] = `
[ZodError: [
  {
    "code": "too_big",
    "maximum": 3,
    "type": "array",
    "inclusive": true,
    "message": "Too many items ðŸ˜¡",
    "path": []
  }
]]
`;

exports[`zod refine and superRefine should be failed 2`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "No duplicated allowed.",
    "path": []
  }
]]
`;

exports[`zod transform should be failed 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Not a number",
    "path": []
  }
]]
`;
